### DISCUSSION TITLE
Simpler declaration of generic type constraints

### DISCUSSION THREAD

**Discussion by ekolis** [dotnet/csharplang#4021] [General] [answered]
In C#, you have to do some repetitive typing in order to declare a generic type constraint:

```
public class MyClass<T> where T : ISomething
```

It's much simpler in VB and Java; how about something like this, inspired by those languages:

```
public class MyClass<T: ISomething>
```

**Comment by alexrp** [comment, marked as answer]
Once you add in non-contrived names, multiple generic parameters and/or constraints, an inheritance list, and perhaps more class modifiers, the current syntax starts to make a lot more sense. Add those things to the Java/VB syntax and it becomes unwieldy fast.

**Reply by ekolis** [reply]
Yeah, I guess you've got a point. Would be nice to have the option to do it either way, but it's not a big deal!

**Reply by huoyaoyuan** [reply]
> but it's not a big deal

Any language feature isn't small. There're IDE support, test coverage, analyzer correctness to work.

> Would be nice to have the option to do it either way

Generally the language team doesn't like to support one thing in different ways.

**Reply by ekolis** [reply]
I understand. I always like to think of Microsoft as the "there's multiple ways to do everything so in case you forget or dislike one there's always another" software company (just look at Settings vs. Control Panel in Windows, and menus vs. toolbar icons in Office, and Ctrl-V to paste vs. Shift-Insert to paste, etc.) but being a software developer myself I can definitely grasp the need for thorough implementation and testing of any new features! Thanks! 🙂
